alphacsc/_solver_d.py

```diff
@@ -0,0 +1,692 @@
+        verbose: int
+            Verbosity level.
+        debug : bool
+            If True, return the cost at each iteration.
+
+        Returns
+        -------
+        uv_hat : array, shape (n_atoms, n_channels + n_times_atom)
+            The atoms to learn from the data.
+        """
+
+        n_channels = z_encoder.n_channels
+
+        uv_hat0, tukey_window_ = self._window(z_encoder.D_hat,
+                                              n_channels,
+                                              z_encoder.n_times_atom)
+
+        objective = self._objective(z_encoder, tukey_window_)
+
+        # use FISTA on alternate u and v
+        adaptive_step_size = (self.solver_d == 'alternate_adaptive')
+
+        uv_hat = uv_hat0.copy()
+        u_hat, v_hat = uv_hat[:, :n_channels], uv_hat[:, n_channels:]
+
+        def prox_u(u, step_size=None):
+            u /= np.maximum(1., np.linalg.norm(u, axis=1, keepdims=True))
+            return u
+
+        def prox_v(v, step_size=None):
+            if self.window:
+                v *= tukey_window_
+
+            v /= np.maximum(1., np.linalg.norm(v, axis=1, keepdims=True))
+
+            if self.window:
+                v /= tukey_window_
+
+            return v
+
+        pobj = []
+        for jj in range(1):
+            # ---------------- update u
+
+            def obj(u):
+                uv = np.c_[u, v_hat]
+                return objective(uv)
+
+            def grad_u(u):
+                if self.window:
+                    uv = np.c_[u, v_hat * tukey_window_]
+                else:
+                    uv = np.c_[u, v_hat]
+
+                grad_d = gradient_d(uv,
+                                    X=z_encoder.X,
+                                    z=z_encoder.get_z_hat(),
+                                    constants=z_encoder.get_constants(),
+                                    loss=z_encoder.loss,
+                                    loss_params=z_encoder.loss_params)
+
+                return (grad_d * uv[:, None, n_channels:]).sum(axis=2)
+
+            if adaptive_step_size:
+                Lu = 1
+            else:
+                if z_encoder.loss != 'l2':
+                    raise NotImplementedError()
+                Lu = self.compute_lipschitz(uv_hat,
+                                            z_encoder.n_channels,
+                                            z_encoder.ztz,
+                                            'u')
+            assert Lu > 0
+
+            u_hat, pobj_u = fista(obj, grad_u, prox_u, 0.99 / Lu, u_hat,
+                                  self.max_iter, momentum=self.momentum,
+                                  eps=self.eps,
+                                  adaptive_step_size=adaptive_step_size,
+                                  debug=debug, verbose=verbose,
+                                  name="Update u")
+            uv_hat = np.c_[u_hat, v_hat]
+            if debug:
+                pobj.extend(pobj_u)
+
```

Comment:
To make this code easier to read, maybe move each update step in separate functions?